src/main/java/com/yahoo/bullet/bql/query/QueryValidator.java

```diff
@@ -0,0 +1,228 @@
+        }
+
+        if (processedQuery.getProjection() != null) {
+            expressionValidator.process(processedQuery.getProjection(), processedQuery.getPreAggregationMapping());
+            List<Schema.Field> fields = toSchemaFields(query.getProjection().getFields());
+            duplicates(fields).ifPresent(duplicates -> {
+                processedQuery.getErrors().add(new BulletError("The following field names/aliases are shared: " + duplicates,
+                                                               "Please specify non-overlapping field names and aliases."));
+            });
+            if (query.getProjection().getType() == Projection.Type.COPY) {
+                schema.addLayer(new Schema(fields));
+            } else {
+                schema.replaceSchema(new Schema(fields));
+            }
+        }
+
+        // Process group by and aggregate nodes. Order doesn't matter. It only matters that they're using the schema after projections.
+        // If the query type is select distinct, the select nodes are the actual aggregate nodes.
+        ProcessedQuery.QueryType queryType = processedQuery.getQueryType();
+
+        if (queryType == ProcessedQuery.QueryType.SELECT_DISTINCT) {
+            List<ExpressionNode> selectItems = processedQuery.getSelectNodes().stream().map(SelectItemNode::getExpression).collect(Collectors.toList());
+            expressionValidator.process(selectItems, processedQuery.getPreAggregationMapping());
+            // Primitives only
+            for (ExpressionNode node : selectItems) {
+                Type type = processedQuery.getPreAggregationMapping().get(node).getType();
+                if (!Type.isUnknown(type) && !Type.isPrimitive(type)) {
+                    processedQuery.getErrors().add(new BulletError("The SELECT DISTINCT field " + node + " is non-primitive. Type given: " + type,
+                                                                   "Please specify primitive fields only for SELECT DISTINCT."));
+                }
+            }
+        } else {
+            expressionValidator.process(processedQuery.getAggregateNodes(), processedQuery.getPreAggregationMapping());
+            expressionValidator.process(processedQuery.getGroupByNodes(), processedQuery.getPreAggregationMapping());
+            // Primitives only
+            for (ExpressionNode node : processedQuery.getGroupByNodes()) {
+                Type type = processedQuery.getPreAggregationMapping().get(node).getType();
+                if (!Type.isUnknown(type) && !Type.isPrimitive(type)) {
+                    processedQuery.getErrors().add(new BulletError("The GROUP BY field " + node + " is non-primitive. Type given: " + type,
+                                                                   "Please specify primitive fields only for GROUP BY."));
+                }
+            }
+
+        }
+
+        // Replace schema with aggregate fields
+        List<Schema.Field> aggregateFields = null;
```

Comment:
Might want to break this method up into smaller chunks for each part of the query?